compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -4041,6 +4050,58 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    fn detect_and_explain_multiple_crate_versions(
+        &self,
+        err: &mut Diag<'_>,
+        item: &ty::AssocItem,
+        owner: hir::OwnerId,
+        rcvr_ty: Ty<'_>,
+    ) -> bool {
+        let pick_name = self.tcx.crate_name(item.def_id.krate);
+        if let Some(map) = self.tcx.in_scope_traits_map(owner) {
+            for trait_candidate in map.to_sorted_stable_ord().into_iter().flat_map(|v| v.1.iter()) {
```

Comment:
This is the thing I'm slightly uneasy about: for *every* error of a method not found that triggers the prior conditions, we'll be doing a linear scan of all traits in scope. Don't have a good sense of whether that will be a significant slow-down for compiles with errors (not a good way of testing that difference).

Reply:
Please take some time to fuzz this. Would it be possible to trigger this in some normal session by removing some imports or upgrading a crate to a newer version (where the trait method is removed in the newer version and the older version is still imported by some transitive dependencies)?

It's very possible that some large project trying to refactor something will hit this if this is a significant slowdown.

Reply:
I think putting this behind a query that allocates a `HashMap<crate name, Vec<trait DefId>>` and returns it would reduce the amortized cost significantly and eliminate the risk of causing trouble.

Reply:
Ok, I changed it to use `in_scope_traits` instead of `in_scope_traits_map`, which will be much cheaper (at the cost of only looking at traits in the current scope, which is fine).

Reply:
@fee1-dead given the current approach (where we're looking at in-scope traits for a *specific* scope), which is something we're already doing for every method resolution, I feel more comfortable with it. I'm also not quite sure how to synthesize a big enough corpus of "wrong crate" to properly stress this.